src/pretix/base/services/cross_selling.py

```diff
@@ -0,0 +1,201 @@
+        returns a queryset of the items that should be displayed, as well as a dict giving additional information on them.
+
+        :returns: (QuerySet<Item>, dict<item_pk: (max_count, discount_rule)>)
+            max_count is `inf` if the item should not be limited
+            discount_rule is None if the item will not be discounted
+        """
+        if category.cross_selling_mode is None:
+            return None, {}
+        if category.cross_selling_condition == 'always':
+            return category.items.all(), {}
+        if category.cross_selling_condition == 'products':
+            match = set(match.pk for match in category.cross_selling_match_products.only('pk'))  # TODO prefetch this
+            return (category.items.all(), {}) if any(pos.item.pk in match for pos in self.cartpositions) else (None, {})
+        if category.cross_selling_condition == 'discounts':
+            my_item_pks = [item.id for item in category.items.all()]
+            potential_discount_items = {
+                item.pk: (max_count, discount_rule)
+                for item, max_count, discount_rule in self._potential_discounts_by_item_for_current_cart
+                if max_count > 0 and item.pk in my_item_pks and item.is_available()
+            }
+
+            return category.items.filter(pk__in=potential_discount_items), potential_discount_items
+
+    @cached_property
+    def _potential_discounts_by_item_for_current_cart(self):
+        potential_discounts_by_cartpos = defaultdict(list)
+
+        from ..services.pricing import apply_discounts
+        self._discounted_prices = apply_discounts(
+            self.event,
+            self.sales_channel,
+            [
+                (cp.item_id, cp.subevent_id, cp.line_price_gross, bool(cp.addon_to), cp.is_bundled,
+                 cp.listed_price - cp.price_after_voucher)
+                for cp in self.cartpositions
+            ],
+            collect_potential_discounts=potential_discounts_by_cartpos
+        )
+
+        # flatten potential_discounts_by_cartpos (a dict of lists of potential discounts) into a set of potential discounts
+        # (which is technically stored as a dict, but we use it as an OrderedSet here)
+        potential_discount_set = dict.fromkeys(
+            info for lst in potential_discounts_by_cartpos.values() for info in lst)
+
+        # sum up the max_counts and pass them on (also pass on the discount_rules so we can calculate actual discounted prices later):
+        # group by benefit product
+        # - max_count for product: sum up max_counts
+        # - discount_rule for product: take first discount_rule
+
+        def discount_info(item, infos_for_item):
+            infos_for_item = list(infos_for_item)
+            return (
+                item,
+                sum(max_count for (item, discount_rule, max_count, i) in infos_for_item),
+                next(discount_rule for (item, discount_rule, max_count, i) in infos_for_item)
+            )
```

Comment:
Wouldn't we need subevent here as well for proper handling in case discount.subevent_mode is set to SUBEVENT_MODE_SAME or am I missing a piece?